Reuse soft-wrap infrastructure in :reflow

rhogenson opened the pull request:
This uses the single-line comment prefixes from the language config, so it should be able to handle different languages in a robust way. The logic is fairly simple, and doesn't handle block comments, for example.

Fixes #3332, #3622

Thanks for all your work on this editor, it is a joy to use.

pascalkuthe commented:
We want to move away from textwrap to our own wrapping system. That will allow us to implement these kind of thjngs without resorting to hacks.

the-mikedavis commented:
In particular see https://github.com/helix-editor/helix/issues/3332#issuecomment-1785731256

rhogenson commented:
Yes, I saw that comment about wanting to have a single implementation for hard and soft wrap, but I think there are a lot of different considerations that apply to hard and soft wrap. This change extends the comment prefix onto subsequent lines, but I don't think that behavior makes sense for soft wrap.

I also don't think it's fair to call this change a "hack," it takes the comment prefix from the language config, so it should be fairly robust

rhogenson commented:
I integrated the reflow logic into DocumentFormatter so that we can reuse the existing soft-wrap logic. This does mean that :reflow always preserves existing newlines, so it no longer allows reflowing out to a larger width, but it also makes it more consistent and probably accidentally addresses #2419

TornaxO7 commented on helix-term/src/commands/typed.rs:
```diff
@@ -2118,14 +2119,35 @@ fn reflow(
+        wrap_indicator_highlight: None,
+        viewport_width: u16::try_from(text_width).unwrap_or(u16::MAX),
+        soft_wrap_at_text_width: true,
+        continue_comments: Vec::from(
+            doc.language_config()
+                .and_then(|config| config.comment_tokens.as_deref())
+                .unwrap_or(&[]),
+        ),
```
Hm... I'm wondering if it's worth it to just provide an `Option<&'a [&str]>` instead of a `Vec<&str>` to avoid the allocation.
I mean the clone isn't that expensive so it should be fine as well.
Just my two cents :eyes:

rhogenson replied:
It's a little inconvenient to have TextFormat borrow from doc. For example, having TextFormat borrow from doc breaks this code: https://github.com/helix-editor/helix/blob/a0bd39d40e983c9fdcd4909b0077b3ac8bf43ad6/helix-term/src/commands.rs#L1762

Since it's also using doc mutably in other places, having text_format borrow doc immutably causes a compile error. It's not obvious to me how to rewrite that code to avoid the issue.

Languages should typically only have 1 or 2 comment prefixes, so I think a small clone is worth it here to avoid lifetime hell.

TornaxO7 replied:
Fair, thank you for the insight :+1:

rhogenson replied:
Fixed in the latest commit

Commit:
Reuse soft-wrap infrastructure in :reflow.

I changed :reflow to use the DocumentFormatter object instead of the
textwrap crate. This allows using the same logic for soft wrap as
for :reflow. Because the logic is the same as for soft wrap, we end up
preserving all existing newlines, so it's more like "wrap" than reflow,
but I think this behavior makes sense anyway to avoid extraneous diffs.

Commit:
Handle single-line comment prefixes in :reflow.

rhogenson commented:
I fixed some bugs in the implementation, but there's still one big issue remaining: it breaks URLs across lines since soft-wrap currently wraps at punctuation. I think that's being tracked as a separate issue though (#11428)

Commit:
Only insert hard-wraps at whitespace.

This is a better behavior for :reflow since it prevents breaking URLs
across lines.

rhogenson commented:
I fixed the remaining issues and resolved the merge conflicts. This change is ready for review

Commit:
Separate reflow from DocumentFormatter

It was making things much more complicated by needing to handle two
different cases in the same code path. By separating reflow out, it can
be more efficient and easier to debug.

Commit:
Remove unnecessary fields

Commit:
Fix grapheme indexing

rhogenson commented:
> What's the status of this? IMO, it's much awaited and needed by a lot of people, as basically atm reflowing a long comment every time is just painful.

I've been running this patch in my fork for months, and the code is fairly simple and well-tested, but I would still appreciate anyone who has a chance to review the code. I'll continue to keep this branch up to date with merge conflicts

the-mikedavis commented:
Ideally I'd like @pascalkuthe to give this a look since he authored most of the soft-wrap code and is most familiar with this section. At a quick glance I think this looks good and I'll give it a proper look after the coming release.

Commit:
Fix wrapping a long word at the start of a line

There's an edge case where any indent at the start of the line shouldn't
be counted as a word boundary.

the-mikedavis reviewed (commented):
To me it looks like `helix_core::doc_formatter::reflow` does not reuse the soft-wrap infrastructure but makes a new greedy reflow based only on grapheme clusters which happens to be in the same module.

I'm split on this change because I think that a greedy reflow is not a bad idea in general. AFAIK Vim also does a greedy reflow. Refactoring the reflow code to consider comment-tokens and other editor-specific knowledge like line ending seems valuable. What I was imagining, though, was something that reuses the document formatter but makes changes to enable breaking lines by inserting characters.

If this is generally an improvement - i.e. makes `:reflow` work on languages where it does not currently - then I'm inclined to merge it even without integrating with the document formatter though, since I think it would be a tangible improvement.

@pascalkuthe what do you think? Would you block this on extending the document formatter?

ThePuzzlemaker commented:
Hello! It has been some time and this PR (or another that fixes the underlying issues) would be very helpful for my workflow. Are there any blocking items on this that need work, or did it just fall under the radar?

Thanks all for your amazing work on this editor!

git-staus commented:
@rhogenson Can I just ask if this PR is ready for review again or if you are still integrating the feedback from the last review?

I am just a passerby who ran into #3332 just now, and I am happy to see this issue is still moving forward. I want to thank all participants in this thread for sharing your hard work with us all!

rhogenson commented:
This PR should be ready for review! Although I wasn't able to completely unify the soft-wrap and reflow implementations, I think it keeps the code simpler and more maintainable this way since these two use cases are actually fairly different.

If the maintainers still feel strongly that these should be done in the same code path, it will need to be a new PR since I wasn't able to get it to work.
